lib/Transforms/Scalar/ScalarReplAggregatesHLSL.cpp

```diff
@@ -6961,7 +6984,7 @@ void ReplaceMultiDimGEP(User *GEP, Value *OneDim, IRBuilder<> &Builder) {
 void MultiDimArrayToOneDimArray::lowerUseWithNewValue(Value *MultiDim, Value *OneDim) {
```

Comment:
Are you testing these changes? I'm not seeing multi-dimensional arrays in the tests.

Reply:
Preserving vectors creates multi-dim arrays because vectors are translated to arrays.  A scenario with multi-dim groupshared array passed through `this`, produces an addrspace cast that survives inlining and was never tested before.  This code change is required to prevent it from asserting in this valid scenario.  As a TODO, we could add a pass to eliminate addrspace casts for shader targets (not libraries), since all the code will be inlined and lead only to loads/stores (through GEPs).